hotham/src/shaders/culling.comp

```diff
@@ -27,48 +27,24 @@ layout(std430, set = 0, binding = 1) writeonly buffer DrawCommandsBuffer {
 } drawCommandsBuffer;
 
 layout(set = 0, binding = 2) uniform CullData {
-    mat4 frustum;
-	uint drawCalls;
+    mat4 clipPlanes1;
+    mat4 clipPlanes2;
+    uint drawCalls;
 } cullData;
 
-bool check_is_visible(mat4 mat, vec3 origin, float radius)
-{
-	uint plane_index = 0;
-	for (uint i = 0; i < 3; ++i)
-	{
-		for (uint j = 0; j < 2; ++j, ++plane_index)
-		{
-			if (plane_index == 2 || plane_index == 3)
-			{
-				continue;
-			}
-			const float sign  = (j > 0) ? 1.f : -1.f;
-			vec4        plane = vec4(0, 0, 0, 0);
-			for (uint k = 0; k < 4; ++k)
-			{
-				plane[k] = mat[k][3] + sign * mat[k][i];
-			}
-			plane.xyzw /= sqrt(dot(plane.xyz, plane.xyz));
-			if (dot(origin, plane.xyz) + plane.w + radius < 0)
-			{
-				return false;
-			}
-		}
-	}
-
-	return true;
-}
-
 void main() {
     uint id = gl_GlobalInvocationID.x;
 
-	if (id >= cullData.drawCalls) { return; }
+    if (id >= cullData.drawCalls) { return; }
 
     DrawData d = drawDataBuffer.data[id];
-    vec3 center = d.boundingSphere.xyz;
-    float radius = d.boundingSphere.w;
-
-	// // Check each eye to see if the object is visible
-    bool isVisible = check_is_visible(cullData.frustum, center, radius);
+    vec4 center4 = vec4(d.boundingSphere.xyz, 1);
+    vec4 negRadius4 = -d.boundingSphere.wwww;
+
+    // Treat as visible if at least one eye may see it
+    bool isVisible = any(bvec2(
+        all(greaterThan(cullData.clipPlanes1 * center4, negRadius4)),
+        all(greaterThan(cullData.clipPlanes2 * center4, negRadius4))
+    ));
     drawCommandsBuffer.drawCommands[id].instanceCount = isVisible ? 1 : 0;
```

Comment:
Minor optimisation: cast the bool to a uint with uint(isVisible) - saves a couple instructions

Reply:
FWIW, there's approximately zero chance the driver won't produce equivalent code for those.

Reply:
On desktop land these assumptions are reasonable, but Qualcomm's documentation constantly warns about not assuming the driver's compiler will do what you want. 😛